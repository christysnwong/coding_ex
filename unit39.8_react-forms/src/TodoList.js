import React, {useState} from "react";
import NewTodoForm from "./NewTodoForm";
import Todo from "./Todo";

const TodoList = () => {
    const INITIAL_STATE = [];
    const [todos, setTodos] = useState(INITIAL_STATE);
    const [todoId, setTodoId] = useState(1);

    const addTodo = (newTodo) => {
        setTodos(todos => [...todos, { ...newTodo, id: todoId}]);
        setTodoId(todoId => todoId + 1);
    }

    const removeTodo = (id) => {
        setTodos(todos => todos.filter( t => t.id !== id));
    }
    

    const updateTodo = (id, updatedTodo) => {
        setTodos(todos =>
            todos.map(todo =>
                todo.id === id? {...todo, task: updatedTodo} : todo
            )    
        );
    }

    return (
        <div>
            <h3>TodoList</h3>
            <NewTodoForm addTodo={addTodo} remove={removeTodo}/>
            <div>
                {todos.map(todo => 
                    <Todo
                        key={todo.id}
                        id={todo.id}
                        task={todo.task}
                        removeTodo={removeTodo}
                        updateTodo={updateTodo}
                    />

                )}
                
            </div>
           
        </div>
    )
}

export default TodoList;