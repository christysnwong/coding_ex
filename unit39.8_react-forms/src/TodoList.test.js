import { render, fireEvent } from "@testing-library/react";
import TodoList from "./TodoList";

function addTodo(list) {
    const input = list.getByLabelText("Task");
    const addBtn = list.getByText("Add");

    fireEvent.change(input, { target: { value: "mop floor" } });
    fireEvent.click(addBtn);
};

it("renders without crashing", function () {
  render(<TodoList />);
});

it("matches snapshot", function () {
  const { asFragment } = render(<TodoList />);
  expect(asFragment()).toMatchSnapshot();
});

it("shows add form correctly", function() {
    const { getByText, getByLabelText } = render(<TodoList />);
    expect(getByText("Add")).toBeInTheDocument();
    expect(getByLabelText("Task")).toBeInTheDocument();
})

it("adds task", function() {
    const list = render(<TodoList />);
    addTodo(list);

    expect(list.getByText("mop floor", {exact: false})).toBeInTheDocument();
    expect(list.getByText("X")).toBeInTheDocument();
    expect(list.getByText("Update")).toBeInTheDocument();
    expect(list.getByText("Mark as completed")).toBeInTheDocument();

});

it("marks task completed", function() {
    const list = render(<TodoList />);
    addTodo(list);

    const spanTask = list.getByText("- mop floor");
    const markBtn = list.getByText("Mark as completed");
    fireEvent.click(markBtn);

    expect(spanTask).toHaveClass("marked");

})


it("removes task", function() {
    const list = render(<TodoList />);
    addTodo(list);

    expect(list.getByText("- mop floor")).toBeInTheDocument();

    const removeBtn = list.getByText("X");
    fireEvent.click(removeBtn);

    expect(list.queryByText("- mop floor")).not.toBeInTheDocument();


})


it("shows edit form and edits correctly", function() {
    const list = render(<TodoList />);
    addTodo(list);

    const updateBtn = list.getByText("Update");
    fireEvent.click(updateBtn);

    const editInput = list.getByLabelText("Change");

    const editBtn = list.getByText("Edit");
    expect(editBtn).toBeInTheDocument();

    fireEvent.change(editInput, { target: { value: "vacuum floor" } });
    fireEvent.click(editBtn);

    expect(list.queryByText("- vacuum floor")).toBeInTheDocument();

})

