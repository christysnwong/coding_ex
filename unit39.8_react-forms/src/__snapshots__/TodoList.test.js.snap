// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches snapshot 1`] = `
<DocumentFragment>
  <div>
    <h3>
      TodoList
    </h3>
    <form>
      <label
        for="task"
      >
        Task 
      </label>
      <input
        id="task"
        name="task"
        placeholder="...type in task description"
        type="text"
        value=""
      />
      <button>
        Add
      </button>
    </form>
    <div />
  </div>
</DocumentFragment>
`;
